components/server/ee/src/workspace/gitpod-server-impl.ts

```diff
@@ -1304,9 +1377,107 @@ export class GitpodServerEEImpl extends GitpodServerImpl {
         return chargebeeSubscriptionId;
     }
 
-    // Team Subscriptions
+    // Team Subscriptions 2
+    async getTeamSubscription(ctx: TraceContext, teamId: string): Promise<TeamSubscription2 | undefined> {
+        this.checkUser("getTeamSubscription");
+        await this.guardTeamOperation(teamId, "get");
+        return this.teamSubscription2DB.findForTeam(teamId, new Date().toISOString());
+    }
+
+    protected async onTeamMemberAdded(userId: string, teamId: string): Promise<void> {
+        const now = new Date();
+        const teamSubscription = await this.teamSubscription2DB.findForTeam(teamId, now.toISOString());
+        if (!teamSubscription) {
+            // No team subscription, nothing to do 🌴
+            return;
+        }
+        await this.updateTeamSubscriptionQuantity(teamSubscription);
+        await this.teamSubscription2Service.addTeamMemberSubscription(teamSubscription, userId);
+    }
+
+    protected async onTeamMemberRemoved(userId: string, teamId: string, teamMembershipId: string): Promise<void> {
+        const now = new Date();
+        const teamSubscription = await this.teamSubscription2DB.findForTeam(teamId, now.toISOString());
+        if (!teamSubscription) {
+            // No team subscription, nothing to do 🌴
+            return;
+        }
+        await this.updateTeamSubscriptionQuantity(teamSubscription);
+        await this.teamSubscription2Service.cancelTeamMemberSubscription(
+            teamSubscription,
+            userId,
+            teamMembershipId,
+            now,
+        );
+    }
+
+    protected async onTeamDeleted(teamId: string): Promise<void> {
+        const now = new Date();
+        const teamSubscription = await this.teamSubscription2DB.findForTeam(teamId, now.toISOString());
+        if (!teamSubscription) {
+            // No team subscription, nothing to do 🌴
+            return;
+        }
+        const chargebeeSubscriptionId = teamSubscription.paymentReference;
+        await this.chargebeeService.cancelSubscription(
+            chargebeeSubscriptionId,
+            {},
+            { teamId, chargebeeSubscriptionId },
+        );
+    }
+
+    protected async updateTeamSubscriptionQuantity(teamSubscription: TeamSubscription2): Promise<void> {
+        const members = await this.teamDB.findMembersByTeam(teamSubscription.teamId);
+        const oldQuantity = teamSubscription.quantity;
+        const newQuantity = members.length;
+        log.info(`updateTeamSubscriptionQuantity: oldQuantity=${oldQuantity}, newQuantity=${newQuantity}`);
+        try {
+            if (oldQuantity < newQuantity) {
```

Comment:
is it worth adding a check for equality?

the whole process is a chain of many asyncs/promises, so it's possible for additions/removals to balance out.

Reply:
Sure, I'll add a check and error log.